oracle/src/lib.rs

```diff
@@ -1,30 +1,108 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use support::{decl_event, decl_module, decl_storage};
+mod default_combine_data;
+mod mock;
+mod operator_provider;
+mod tests;
+mod timestamped_value;
+
+pub use default_combine_data::DefaultCombineData;
+pub use operator_provider::OperatorProvider;
+use rstd::prelude::Vec;
+use rstd::result;
+use sr_primitives::traits::Member;
+use support::{decl_error, decl_event, decl_module, decl_storage, dispatch::Result, ensure, traits::Time, Parameter};
+use system::ensure_signed;
+pub use timestamped_value::TimestampedValue;
+pub use traits::{CombineData, OnNewData};
+
+type MomentOf<T> = <<T as Trait>::Time as Time>::Moment;
 
 pub trait Trait: system::Trait {
 	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+	type OnNewData: OnNewData<Self::Key, Self::Value>;
+	type OperatorProvider: OperatorProvider<Self::AccountId>;
+	type CombineData: CombineData<Self::Key, TimestampedValue<Self::Value, MomentOf<Self>>>;
+	type Time: Time;
+	type Key: Parameter + Member + Copy + Ord;
+	type Value: Parameter + Member + Copy + Ord;
```

Comment:
For `Key` and `Value`, I feel like we could keep `Key` having a `Copy` bound, but `Value` shouldn't have `Copy` traits since it might have big size. Sounds reasonable @xlc ?